File: Spikes/ProTrinketLSM/Effects.cpp
// 
// 
// 

#ifndef _EFFECTS_h
	#include "Effects.h"
#endif

#include <string.h>

CommaDelimitReturn Effects::cdrPixel, Effects::cdrColor;

void Effects::allClear(Adafruit_DotStar* strip, LightingSequence* lseqs, uint16_t currentSequence, Strip* PassedStripClass) {
	//Exit if invalid pointers passed
	if (strip == NULL) {
		return;
	}
	else if (lseqs == NULL) 
	{
		return;
	}

	//Exit if pixel colors is null
	if (lseqs[currentSequence].colors == NULL) {
		return;
	}

	//Loop through all pixels and set color to clear
	for (int i = 0; i < lseqs[currentSequence].totalPixels; i++) {
		setSinglePixelColor(strip, lseqs, currentSequence, i, 0);
	}

	//Show new effects
	strip->show();
}

//void Effects::rainbow(Adafruit_DotStar* strip, LightingSequence* lseqs, uint16_t currentSequence, int16_t* i, int16_t* p0, int16_t* p1, int16_t* p2, int16_t* p3, int16_t* p4, int16_t* p5, Strip* PassedStripClass) {
//	//Exit if invalid pointers passed
//	if (strip == NULL) {
//		return;
//	}
//	else if (lseqs == NULL)
//	{
//		return;
//	}
//
//	//Add 1 to i if i is -1 and reset ps
//	if (*i < 0) {
//		//Add 1 to i making i = 0
//		*i = *i + 1;
//
//		//Reset p values
//		*p0 = -1, *p1 = 0, *p2 = 1, *p3 = 2, *p4 = 3, *p5 = 4;
//	}
//
//	//Verify if p0 is greater than numPixels, and if so add 1 to shiftPixelsBy
//	if (*p0 >= lseqs[currentSequence].totalPixels) {
//		//Add 1 to shiftPixelsBy and reset j
//		*i = *i + 1;
//
//		//Reset p values
//		*p0 = 0, *p1 = 1, *p2 = 2, *p3 = 3, *p4 = 4, *p5 = 5;
//	}
//
//	if (*i <= lseqs[currentSequence].iterations) {
//		strip->setPixelColor((*p0)++, CLEAR);
//		strip->setPixelColor((*p1)++, RED);
//		strip->setPixelColor((*p2)++, ORANGE);
//		strip->setPixelColor((*p3)++, YELLOW);
//		strip->setPixelColor((*p4)++, GREEN);
//		strip->setPixelColor((*p5)++, BLUE);
//		strip->setPixelColor(*p1 - 2, CLEAR);
//
//		if (*p0 >= lseqs[currentSequence].totalPixels)
//		{
//			*p0 = 0;
//		}
//		if (*p1 >= lseqs[currentSequence].totalPixels)
//		{
//			*p1 = 0;
//		}
//		if (*p2 >= lseqs[currentSequence].totalPixels)
//		{
//			*p2 = 0;
//		}
//		if (*p3 >= lseqs[currentSequence].totalPixels)
//		{
//			*p3 = 0;
//		}
//		if (*p4 >= lseqs[currentSequence].totalPixels)
//		{
//			*p4 = 0;
//		}
//		if (*p5 >= lseqs[currentSequence].totalPixels)
//		{
//			*p5 = 0;
//		}
//	}
//
//	strip->show();
//}

void Effects::loadColor(Adafruit_DotStar* strip, LightingSequence* lseqs, uint16_t currentSequence, uint16_t shiftPixelToLoad, Strip* PassedStripClass) {
	//Exit if invalid pointers passed
	if (strip == NULL) {
		return;
	}
	else if (lseqs == NULL)
	{
		return;
	}

	//Exit if pixelElems or colorElems is null
	if (lseqs[currentSequence].colors == NULL) {
		return;
	}

	//Declare variables
	int elem = 0, colorValue = 0;//, tot = lseqs[currentSequence].totalPixels;
	
	//Reset nextBegPositions then continue
	cdrColor.nextBegPosition = 0;

	for (elem = 0; elem < lseqs[currentSequence].totalPixels; elem++) {
		//Set nextBegPoistion for colors and get next color
		getNextCommaDelimitedColorToken(lseqs[currentSequence].colors, cdrColor.nextBegPosition);

		//Convert return color index to actual color
		setSinglePixelColor(strip, lseqs, currentSequence, (elem + shiftPixelToLoad), cdrColor.value);
	}

	//Show new effects
	strip->show();
}

void Effects::bounceBack(Adafruit_DotStar* strip, LightingSequence* lseqs, uint16_t currentSequence, bool* init, bool* forward, int16_t *shiftPixelsBy, int16_t* tail, int16_t* head, uint16_t* bounces, uint16_t initHead, uint16_t initTail, Strip* PassedStripClass) {
	//Exit if invalid pointers passed
	if (strip == NULL) {
		return;
	}
	else if (lseqs == NULL)
	{
		return;
	}

	//Initilialize if first time in
	if (*init) {
		//Simply set colors, setup values, and exit
		loadColor(strip, lseqs, currentSequence, 0, PassedStripClass);

		*init = false;
		*bounces = lseqs[currentSequence].bounces;
		*shiftPixelsBy = 0;
		*head = initHead;
		*tail = initTail;

		return;
	}

	if (*bounces > 0) {
		if (*forward) {
			//Add 1 to shiftPixelsBy
			*shiftPixelsBy = *shiftPixelsBy + 1;
			*head = *head + 1;
			*tail = *tail + 1;

			if (*head >= lseqs[currentSequence].totalPixels)
			{
				*forward = false;
				*bounces = *bounces + 1;
				*shiftPixelsBy = *shiftPixelsBy - 1;
				*head = *head - 1;
				*tail = *tail - 1;
			}

			//Load pixel colors
			loadColor(strip, lseqs, currentSequence, *shiftPixelsBy, PassedStripClass);
		}
		else {
			//Subtract 1 from shiftPixelsBy
			*shiftPixelsBy = *shiftPixelsBy - 1;
			*head = *head - 1;
			*tail = *tail - 1;

			if (*tail < 0)
			{
				*forward = true;
				*bounces = *bounces + 1;
				*shiftPixelsBy = *shiftPixelsBy + 1;
				*head = *head + 1;
				*tail = *tail + 1;
			}

			//Load pixel colors
			loadColor(strip, lseqs, currentSequence, *shiftPixelsBy, PassedStripClass);
		}
	}

	//Show effect on led
	strip->show();
}

void Effects::flowThrough(Adafruit_DotStar* strip, LightingSequence* lseqs, uint16_t currentSequence, int16_t* i, bool isRainbow, int16_t* p0, int16_t* p1, int16_t* p2, int16_t* p3, int16_t* p4, int16_t* p5, int16_t* virtualPixelIndexArray, Strip* PassedStripClass) {
	//Exit if invalid pointers passed
	if (strip == NULL) {
		return;
	}
	else if (lseqs == NULL)
	{
		return;
	}

	//Declare variables
	int elem = 0, colorValue = 0;

	//Add 1 to i if i is -1 and reset ps
	if (*i < 0) {
		//Add 1 to i making i = 0
		*i = *i + 1;

		if (isRainbow) {
			//Reset p values
			*p0 = -1, *p1 = 0, *p2 = 1, *p3 = 2, *p4 = 3, *p5 = 4;
		}
		//else {
		//	//Reset nextBegPositions then continue
		//	cdrColor.nextBegPosition = 0;

		//	//Set virual pixel elements to default element indexes
		//	for (elem = 0; elem < lseqs[currentSequence].totalPixels; elem++) {
		//		//Add 1 to pixel's shift value
		//		virtualPixelIndexArray[elem] = elem - 1;
		//	}
		//}
	}

	if (isRainbow) {
		//Verify if p0 is greater than numPixels, and if so add 1 to shiftPixelsBy
		if (*p0 >= lseqs[currentSequence].totalPixels) {
			//Add 1 to shiftPixelsBy and reset j
			*i = *i + 1;

			//Reset p values
			//*p0 = -1, *p1 = 0, *p2 = 1, *p3 = 2, *p4 = 3, *p5 = 4;
		}

		if (*i <= lseqs[currentSequence].iterations) {
			strip->setPixelColor((*p0)++, CLEAR);
			strip->setPixelColor((*p1)++, RED);
			strip->setPixelColor((*p2)++, ORANGE);
			strip->setPixelColor((*p3)++, YELLOW);
			strip->setPixelColor((*p4)++, GREEN);
			strip->setPixelColor((*p5)++, BLUE);
			//strip->setPixelColor(*p1 - 2, CLEAR);

			if (*p0 >= lseqs[currentSequence].totalPixels)
			{
				*p0 = 0;
			}
			if (*p1 >= lseqs[currentSequence].totalPixels)
			{
				*p1 = 0;
			}
			if (*p2 >= lseqs[currentSequence].totalPixels)
			{
				*p2 = 0;
			}
			if (*p3 >= lseqs[currentSequence].totalPixels)
			{
				*p3 = 0;
			}
			if (*p4 >= lseqs[currentSequence].totalPixels)
			{
				*p4 = 0;
			}
			if (*p5 >= lseqs[currentSequence].totalPixels)
			{
				*p5 = 0;
			}
		}
	}
	else {
		//Verify if first element is greater than total pixels, and if so add 1 to shiftPixelsBy
		if (virtualPixelIndexArray[0] >= lseqs[currentSequence].totalPixels) {
			//Add 1 to shiftPixelsBy and reset j
			*i = *i + 1;
		}

		if (*i <= lseqs[currentSequence].iterations) {
			//Reset nextBegPositions then continue
			cdrColor.nextBegPosition = 0;

			for (elem = 0; elem < lseqs[currentSequence].totalPixels; elem++) {
				//Add 1 to pixel's shift value
				virtualPixelIndexArray[elem] = virtualPixelIndexArray[elem] + 1;

				Serial.print("Pixel elem: ");
				Serial.print(elem);
				Serial.print(" is Virtual Pixel elem: ");
				Serial.println(virtualPixelIndexArray[elem]);
				Serial.println();

				//Verify pixel shift is not over end of led strip
				if (virtualPixelIndexArray[elem] >= lseqs[currentSequence].totalPixels) {
					//Change virtual index to 0
					virtualPixelIndexArray[elem] = 0;
				}

				//Set nextBegPoistion for colors and get next color
				getNextCommaDelimitedColorToken(lseqs[currentSequence].colors, cdrColor.nextBegPosition);

				Serial.print("Pixel elem: ");
				Serial.print(elem);
				Serial.print(" is Virtual Pixel elem: ");
				Serial.print(virtualPixelIndexArray[elem]);
				Serial.print(" and the color value is ");
				Serial.println(cdrColor.value);
				Serial.println();

				//Set pixel color
				setSinglePixelColor(strip, lseqs, currentSequence, virtualPixelIndexArray[elem], cdrColor.value);
			}
		}
		else {
			allClear(strip, lseqs, currentSequence, PassedStripClass);
		}
	}

	strip->show();
}

/*
*/
void Effects::setSinglePixelColor(Adafruit_DotStar* strip, LightingSequence* lseqs, uint16_t currentSequence, uint16_t pixelElem, uint16_t color) {
	//Exit if invalid pointers passed
	if (strip == NULL) {
		return;
	}
	else if (lseqs == NULL)
	{
		return;
	}

	//Convert return color index to actual color
	switch (color)
	{
	case 0:
		strip->setPixelColor(pixelElem, CLEAR);
		break;
	case 1:
		strip->setPixelColor(pixelElem, WHITE);
		break;
	case 2:
		strip->setPixelColor(pixelElem, RED);
		break;
	case 3:
		strip->setPixelColor(pixelElem, GREEN);
		break;
	case 4:
		strip->setPixelColor(pixelElem, BLUE);
		break;
	case 5:
		strip->setPixelColor(pixelElem, YELLOW);
		break;
	case 6:
		strip->setPixelColor(pixelElem, CYAN);
		break;
	case 7:
		strip->setPixelColor(pixelElem, MAGENTA);
		break;
	case 8:
		strip->setPixelColor(pixelElem, ORANGE);
		break;
	default:
		strip->setPixelColor(pixelElem, CLEAR);
		break;
	}
}

/*
*/
void Effects::getNextCommaDelimitedColorToken(const char* String, uint16_t BegPosition) {
	//Declare variables
	CommaDelimitReturn iret;
	char charAccum[5] = { '\0', '\0', '\0', '\0', '\0' };
	uint16_t elem = -1;

	//Set default return values
	iret.nextBegPosition = -1;
	iret.value = -1;

	//Loop through all pixel elements numbers
	for (int i = BegPosition; i < strlen(String); i++)
	{
		//If we reach a comma, turn accumulated chars into a number and set that pixel to clear
		if (String[i] == ',') {
			//Set number value to next beginning
			iret.nextBegPosition = ++i;
			iret.value = atoi(charAccum);

			break;
		}
		else {
			//Verify if this will be last iteration
			if (i == (strlen(String) - 1)) {
				//Add 1 to elem
				elem++;

				//Add char to position in numAccum array
				charAccum[elem] = String[i];

				//Set number value to next beginning
				iret.nextBegPosition = i + 1;
				iret.value = atoi(charAccum);

				break;
			}
			else {
				//Continue accumulating chars to String

				//Add 1 to elem
				elem++;

				//Add char to position in numAccum array
				charAccum[elem] = String[i];
			}
		}
	}

	Effects::cdrColor.nextBegPosition = iret.nextBegPosition;
	Effects::cdrColor.value = iret.value;
}

File: Src/LEDLightingComposer/LEDLightingComposer/Effects.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using System.Windows.Forms;

namespace LEDLightingComposer
{
    public static class Effects
    {
        //Colors
        public const byte CLEAR = 0;
        public const byte WHITE = 1;
        public const byte RED = 2;
        public const byte GREEN = 3;
        public const byte BLUE = 4;
        public const byte YELLOW = 5;
        public const byte CYAN = 6;
        public const byte MAGENTA = 7;
        public const byte ORANGE = 8;

        //Lighting Effect Types
        public const short FILLER = -1;
        public const byte ALLCLEAR = 0;
        public const byte RAINBOW = 1;
        public const byte LOADCOLOR = 2;
        public const byte BOUNCEBACK = 3;
        public const byte FLOWTHROUGH = 4;

        #region Public Methods

        /*
            Function:

            Parameters:

            Returns:
        */
        public static bool allClear(Strip strip, DrawingManager DrawManager)
        {
            //Declare variables
            DrawableObject dbo = null;
            bool bRet = false;
            int i = 0, currentSequence = strip.CurrentSequence;
            Structs.LightingSequence seq;

            try
            {
                seq = strip.LSeqs[currentSequence];
            }catch(Exception ex)
            {
                try
                {
                    //CurrentSequence possibly past all sequences so try last sequence
                    seq = strip.LSeqs[(currentSequence-1)];
                }
                catch (Exception ex2)
                {
                    return bRet;
                }
            }

            //Find drawable object that matches this strip's pin setup for manipulating LEDs onscreen
            dbo = DrawManager.getDrawableObject("PINSETUP", strip.PinSetup.ToString());

            if(dbo == null)
            {
                return bRet;
            }

            //Set all of Drawable Object's LEDS to clear
            for (i = 0; i < seq.totalPixels; i++)
            {
                dbo.Leds[i].LEDColor = Color.Transparent;
            }

            return bRet;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static bool loadColor(Strip strip, DrawingManager DrawManager)
        {
            //Declare variables
            bool bRet = false;
            DrawableObject dbo = null;
            int i = 0, currentSequence = strip.CurrentSequence;
            Structs.LightingSequence seq;

            try
            {
                seq = strip.LSeqs[currentSequence];
            }
            catch (Exception ex)
            {
                return bRet;
            }

            //Exit if invalid pointers passed
            if (strip == null)
            {
                return bRet;
            }

            //Find drawable object that matches this strip's pin setup for manipulating LEDs onscreen
            dbo = DrawManager.getDrawableObject("PINSETUP", strip.PinSetup.ToString());

            if (dbo == null)
            {
                return bRet;
            }

            //Clear all pixels first
            allClear(strip, DrawManager);

            //Set all of Drawable Object's LEDS to clear
            for (i = 0; i < seq.totalPixels; i++)
            {
                try
                {
                    if ((i + strip.ShiftPixelBy) >= seq.totalPixels)
                    {
                        //Attempt to wrap shifted color to beginning of strip of flowthrough effect
                        if (seq.lightsequence == Effects.FLOWTHROUGH)
                        {
                            dbo.Leds[((i + strip.ShiftPixelBy) % seq.totalPixels)].LEDColor = getColorFromCode(int.Parse(seq.colors[i].Split('-')[0].Trim()), seq.brightness);
                        }else
                        {
                            //Do nothing, leave pixels as is
                        }
                    }
                    else if((i + strip.ShiftPixelBy) < 0)
                    {
                        //Do nothing as we don't want to wrap negative numbers back to end, as this would be from the bounce back effect (all effects move forward)

                    }else
                    {
                        dbo.Leds[(i + strip.ShiftPixelBy)].LEDColor = getColorFromCode(int.Parse(seq.colors[i].Split('-')[0].Trim()), seq.brightness);
                    }
                }catch(Exception ex)
                {
                    
                }
            }

            return bRet;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static bool bounceBack(Strip strip, DrawingManager DrawManager, int initHead, int initTail)
        {
            //Declare variables
            bool bRet = false;
            DrawableObject dbo = null;
            int i = 0, currentSequence = strip.CurrentSequence;
            Structs.LightingSequence seq;

            try
            {
                seq = strip.LSeqs[currentSequence];
            }
            catch (Exception ex)
            {
                return bRet;
            }

            //Exit if invalid pointers passed
            if (strip == null)
            {
                return bRet;
            }

            try
            {
                //Clear all pixels first
                //Effects.allClear(strip, DrawManager);

                //Initilialize if first time in
                if (strip.Init)
                {
                    strip.Init = false;
                    strip.Bounces = seq.bounces;
                    strip.ShiftPixelBy = 0;
                    strip.Head = initHead;
                    strip.Tail = initTail;

                    //Simply set colors, setup values, and exit
                    loadColor(strip, DrawManager);

                    bRet = true;

                    return bRet;
                }

                if (strip.Bounces > 0)
                {
                    if (strip.Forward)
                    {
                        //Add 1 to shiftPixelsBy
                        strip.ShiftPixelBy++;
                        strip.Head++;
                        strip.Tail++;

                        if (strip.Head >= seq.totalPixels)
                        {
                            strip.Forward = false;
                            strip.Bounces++;

                            strip.ShiftPixelBy--;
                            strip.Head--;
                            strip.Tail--;
                        }

                        //Load color with a shift and clear previous tail pixel
                        loadColor(strip, DrawManager);
                    }
                    else
                    {
                        //Subtract 1 from shiftPixelsBy
                        strip.ShiftPixelBy--;
                        strip.Head--;
                        strip.Tail--;

                        if (strip.Tail < 0)
                        {
                            strip.Forward = true;
                            strip.Bounces++;

                            strip.ShiftPixelBy++;
                            strip.Head++;
                            strip.Tail++;
                        }

                        //Load color with a shift and clear previous head pixel
                        loadColor(strip, DrawManager);
                    }
                }
                //Set return value to true
                bRet = true;
            }catch(Exception ex)
            {
                
            }

            return bRet;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static bool flowThrough(Strip strip, DrawingManager DrawManager)
        {
            //Declare variables
            bool bRet = false;
            DrawableObject dbo = null;
            int i = 0, currentSequence = strip.CurrentSequence;
            Structs.LightingSequence seq;

            try
            {
                seq = strip.LSeqs[currentSequence];
            }
            catch (Exception ex)
            {
                return bRet;
            }

            //Exit if invalid pointers passed
            if (strip == null)
            {
                return bRet;
            }

            //Declare variables
            int elem = 0;

            //Add 1 to i if i is -1 and reset ps
            if (strip.I < 0)
            {
                //Add 1 to i making i = 0
                strip.I++;

                //Reset shiftpixelsby
                strip.ShiftPixelBy = 0;

                if (strip.IsRainbow)
                {
                    //Reset p values
                    strip.P0 = 0;
                    strip.P1 = 1;
                    strip.P2 = 2;
                    strip.P3 = 3;
                    strip.P4 = 4;
                    strip.P5 = 5;
                }
                else
                {
                    //Set virual pixel elements to default element indexes
                    for (elem = 0; elem < seq.totalPixels; elem++)
                    {
                        //Add 1 to pixel's shift value
                        strip.VirtualPixelArray[elem] = elem - 1;
                    }
                }
            }

            if (strip.IsRainbow)
            {
                //Verify if p0 is greater than numPixels, and if so add 1 to shiftPixelsBy
                if (strip.P0 >= seq.totalPixels)
                {
                    //Add 1 to shiftPixelsBy and reset j
                    strip.I++;

                    //Reset p values
                    strip.P0 = 0; strip.P1 = 1; strip.P2 = 2; strip.P3 = 3; strip.P4 = 4; strip.P5 = 5;
                }

                if (strip.I < seq.iterations)
                {
                    //Find drawable object that matches this strip's pin setup for manipulating LEDs onscreen
                    dbo = DrawManager.getDrawableObject("PINSETUP", strip.PinSetup.ToString());

                    if (dbo == null)
                    {
                        return bRet;
                    }

                    //Set all of Drawable Object's LEDS to clear
                    //allClear(strip, DrawManager);

                    //Set specific pixels to rainbow colors
                    try { if (strip.P0 >= seq.totalPixels) { strip.P0++; } else { dbo.Leds[strip.P0++].LEDColor = Effects.getColorFromCode(CLEAR, seq.brightness); } } catch (Exception ex) { }
                    try { if (strip.P1 >= seq.totalPixels) { strip.P1++; } else { dbo.Leds[strip.P1++].LEDColor = Effects.getColorFromCode(RED, seq.brightness); } } catch (Exception ex) { }
                    try { if (strip.P2 >= seq.totalPixels) { strip.P2++; } else { dbo.Leds[strip.P2++].LEDColor = Effects.getColorFromCode(ORANGE, seq.brightness); } } catch (Exception ex) { }
                    try { if (strip.P3 >= seq.totalPixels) { strip.P3++; } else { dbo.Leds[strip.P3++].LEDColor = Effects.getColorFromCode(YELLOW, seq.brightness); } } catch (Exception ex) { }
                    try { if (strip.P4 >= seq.totalPixels) { strip.P4++; } else { dbo.Leds[strip.P4++].LEDColor = Effects.getColorFromCode(GREEN, seq.brightness); } } catch (Exception ex) { }
                    try { if (strip.P5 >= seq.totalPixels) { strip.P5++; } else { dbo.Leds[strip.P5++].LEDColor = Effects.getColorFromCode(BLUE, seq.brightness); } } catch (Exception ex) { }

                    if (strip.P0 >= seq.totalPixels)
                    {
                        strip.P0 = 0;
                    }
                    if (strip.P1 >= seq.totalPixels)
                    {
                        strip.P1 = 0;
                    }
                    if (strip.P2 >= seq.totalPixels)
                    {
                        strip.P2 = 0;
                    }
                    if (strip.P3 >= seq.totalPixels)
                    {
                        strip.P3 = 0;
                    }
                    if (strip.P4 >= seq.totalPixels)
                    {
                        strip.P4 = 0;
                    }
                    if (strip.P5 >= seq.totalPixels)
                    {
                        strip.P5 = 0;
                    }
                }
            }
            else
            {
                //Verify if first element is greater than total pixels, and if so add 1 to shiftPixelsBy
                if (strip.VirtualPixelArray[0] >= seq.totalPixels)
                {
                    //Add 1 to shiftPixelsBy and reset j
                    strip.I++;
                }

                if (strip.I <= seq.iterations)
                {
                    //Find drawable object that matches this strip's pin setup for manipulating LEDs onscreen
                    dbo = DrawManager.getDrawableObject("PINSETUP", strip.PinSetup.ToString());

                    if (dbo == null)
                    {
                        return bRet;
                    }

                    for (elem = 0; elem < seq.totalPixels; elem++)
                    {
                        //Add 1 to pixel's shift value
                        strip.VirtualPixelArray[elem] = strip.VirtualPixelArray[elem] + 1;

                        //Verify pixel shift is not over end of led strip
                        if (strip.VirtualPixelArray[elem] >= seq.totalPixels)
                        {
                            //Change virtual index to 0
                            strip.VirtualPixelArray[elem] = 0;
                        }

                        //Set pixel color
                        try { dbo.Leds[strip.VirtualPixelArray[elem]].LEDColor = Effects.getColorFromCode(int.Parse(seq.colors[elem].Split('-')[0].Trim()), seq.brightness); } catch (Exception ex) { }
                    }
                }
                else
                {
                    allClear(strip, DrawManager);
                }
            }

            return bRet;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static bool updateBrightness(Strip strip, DrawingManager DrawManager)
        {
            //Declare variables
            bool bRet = false;
            DrawableObject dbo = null;
            int i = 0, currentSequence = strip.CurrentSequence;
            Structs.LightingSequence seq;

            try
            {
                seq = strip.LSeqs[currentSequence];
            }
            catch (Exception ex)
            {
                return bRet;
            }

            //Exit if invalid pointers passed
            if (strip == null)
            {
                return bRet;
            }

            //Find drawable object that matches this strip's pin setup for manipulating LEDs onscreen
            dbo = DrawManager.getDrawableObject("PINSETUP", strip.PinSetup.ToString());

            if (dbo == null)
            {
                return bRet;
            }

            //Update brightness if necessary
            if (strip.IncrementBrightness)
            {
                //Add incr to brightness
                seq.brightness += seq.incrBrightness;

                if (seq.brightness > 255)
                {
                    seq.brightness = 255;
                    seq.incrBrightness *= -1;
                }
                else if (seq.brightness < 0)
                {
                    seq.brightness = 0;
                    seq.incrBrightness *= -1;
                }

                //Set brightness for all of Drawable Object's LEDS to specified brightness
                for (i = 0; i < seq.totalPixels; i++)
                {
                    try
                    {
                        if ((i + strip.ShiftPixelBy) < seq.totalPixels && (i + strip.ShiftPixelBy) >= 0)
                        {
                            dbo.Leds[(i + strip.ShiftPixelBy)].updateBrightness(seq.brightness);
                        }
                    }
                    catch (Exception ex)
                    {

                    }
                }
            }
            else
            {
                //Set brightness for all of Drawable Object's LEDS to specified brightness
                for (i = 0; i < seq.totalPixels; i++)
                {
                    try
                    {
                        if ((i + strip.ShiftPixelBy) < seq.totalPixels && (i + strip.ShiftPixelBy) >= 0)
                        {
                            dbo.Leds[(i + strip.ShiftPixelBy)].updateBrightness(seq.brightness);
                        }
                    }
                    catch (Exception ex)
                    {

                    }
                }
            }

            //Change original struct to match updated temp struct
            strip.LSeqs[currentSequence] = seq;

            bRet = true;

            return bRet;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static Color getColorFromCode(int ColorCode, int Brightness)
        {
            Color clr = Color.Transparent;

            switch (ColorCode)
            {
                case CLEAR:
                    clr = Color.Transparent;
                    break;
                case WHITE:
                    clr = Color.White;
                    break;
                case RED:
                    clr = Color.Red;
                    break;
                case GREEN:
                    clr = Color.Green;
                    break;
                case BLUE:
                    clr = Color.Blue;
                    break;
                case YELLOW:
                    clr = Color.Yellow;
                    break;
                case CYAN:
                    clr = Color.Cyan;
                    break;
                case MAGENTA:
                    clr = Color.Magenta;
                    break;
                case ORANGE:
                    clr = Color.Orange;
                    break;
            }

            clr = updateBrightness(clr, Brightness);

            return clr;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static String getEffectFromCode(int EffectCode)
        {
            String sret = "";

            switch (EffectCode)
            {
                case 0:
                    sret = "ALLCLEAR";
                    break;
                case 1:
                    sret = "RAINBOW";
                    break;
                case 2:
                    sret = "LOADCOLOR";
                    break;
                case 3:
                    sret = "BOUNCEBACK";
                    break;
                case 4:
                    sret = "FLOWTHROUGH";
                    break;
                default:
                    sret = "ALLCLEAR";
                    break;
            }

            return sret;
        }

        /*
            Function:

            Parameters:

            Returns:
        */
        public static Color updateBrightness(Color CLR, int Brightness)
        {
            Color clr = CLR;

            try
            {
                float correctionFactor = ((float)Brightness / (float)255);
                int red = ((int)(clr.R * correctionFactor) > 255) ? 255 : (int)(clr.R * correctionFactor);
                int green = ((int)(clr.G * correctionFactor) > 255) ? 255 : (int)(clr.G * correctionFactor);
                int blue = ((int)(clr.B * correctionFactor) > 255) ? 255 : (int)(clr.B * correctionFactor);
                clr = Color.FromArgb(clr.A, red, green, blue);
            }
            catch(Exception ex)
            {

            }

            return clr;
        }


        #endregion Public Methods
    }
}
